zkevm-circuits/src/taiko_pi_circuit_.rs

```diff
@@ -0,0 +1,832 @@
+
+    block_table: BlockTable,
+    keccak_table: KeccakTable,
+    byte_table: ByteTable,
+
+    // To annotate columns at assignment for debug purpose 
+    col_configs: Vec<CellColumn<F, PiCellType>>,
+}
+
+/// Circuit configuration arguments
+pub struct TaikoPiCircuitConfigArgs<F: Field> {
+    /// BlockTable
+    pub block_table: BlockTable,
+    /// KeccakTable
+    pub keccak_table: KeccakTable,
+    /// ByteTable
+    pub byte_table: ByteTable,
+    /// Challenges
+    pub challenges: Challenges<Expression<F>>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum PiCellType {
+    Storage1,
+    Storage2,
+    Byte,
+    Lookup(Table)
+}
+impl CellType for PiCellType {
+    fn byte_type() -> Option<Self> {
+        Some(Self::Byte)
+    }
+    fn storage_for_phase(phase: u8) -> Self {
+        match phase {
+            1 => PiCellType::Storage1,
+            2 => PiCellType::Storage2,
+            _ => unimplemented!()
+        }
+    }
+}
+impl Default for PiCellType {
+    fn default() -> Self {
+        Self::Storage1
+    }
+}
+
+
+impl<F: Field> SubCircuitConfig<F> for TaikoPiCircuitConfig<F> {
+    type ConfigArgs = TaikoPiCircuitConfigArgs<F>;
+
+        /// Return a new TaikoPiCircuitConfig
+        fn new(
+            meta: &mut ConstraintSystem<F>,
+            Self::ConfigArgs {
+                block_table,
+                keccak_table,
+                byte_table,
+                challenges,
+            }: Self::ConfigArgs,
+        ) -> Self {
+            let field_cm = CellManager::new(
+                meta,
+                vec![
+                    (PiCellType::Byte, 32, 1, false),
+                    (PiCellType::Storage1, 3, 1, false)
+                ],
+                0,
+                1,
+            );
+            let state_cm = CellManager::new(
+                meta,
+                vec![
+                    (PiCellType::Storage1, 1, 1, false),
+                    (PiCellType::Storage2, 1, 2, true),
+                ],
+                0,
+            12,
+            );
+            let mut cb: ConstraintBuilder<F, PiCellType> = ConstraintBuilder::new(4,  Some(field_cm.clone()), Some(challenges.evm_word()));
+            cb.preload_tables(meta,
+                &[
+                   (PiCellType::Lookup(Keccak), &keccak_table), 
+                   (PiCellType::Lookup(Bytecode), &byte_table), 
+                   (PiCellType::Lookup(Block), &block_table)
+                   ]
+               );
+            let q_enable = meta.complex_selector();
+            let public_input = meta.instance_column();
+            let block_acc = meta.advice_column_in(SecondPhase);
+            let pi_field = FieldBytesGadget::config(&mut cb, &challenges);
+            cb.set_cell_manager(state_cm.clone());
+            let block_number = [();2].map(|_| cb.query_one(PiCellType::Storage1));
+            let keccak_output = [();2].map(|_| cb.query_one(PiCellType::Storage2));
```

Comment:
Doe these need to be stored in a separate CellManager? Seems like these could just be stored somewhere in the main CellManager and it should still be okay? `keccak_output` requires a copy constraint I think which should be possible I think, `block_number` doesn't seem to require anything special, just some storage cell, unless I miss something.